tt_eager/tt_dnn/op_library/sdpa/multi_core/sdpa_decode_op_multi_core.cpp

```diff
@@ -0,0 +1,628 @@
+        all_writer_kernels_id.push_back(writer_kernels_id);
+        all_compute_kernels_id.push_back(compute_kernels_id);
+    }
+
+    uint32_t q_addr = q_buffer->address();
+    uint32_t k_addr = k_buffer->address();
+    uint32_t v_addr = v_buffer->address();
+    uint32_t mask_addr = mask_buffer->address();
+    uint32_t out_addr = out0_buffer->address();
+
+
+    // Set reader rt args
+    for (uint32_t i = 0; i < num_active_cores; ++i) {
+        CoreCoord core = core_group[i];
+        uint32_t worker_id = i % num_cores_per_batch - 1;
+        bool do_reduce = (worker_id == -1);
+
+        uintptr_t reader_kernels_id = all_reader_kernels_id[i];
+        uintptr_t writer_kernels_id = all_writer_kernels_id[i];
+        uintptr_t compute_kernels_id = all_compute_kernels_id[i];
+
+        SetRuntimeArgs(program, reader_kernels_id, core, { q_addr, k_addr, v_addr, mask_addr });
+        if (do_reduce) {
+            SetRuntimeArgs(program, writer_kernels_id, core, { out_addr });
+        }
+    }
+
+    auto override_runtime_arguments_callback = [
+        num_active_cores,
+        core_group,
+        all_reader_kernels_id,
+        all_writer_kernels_id,
+        all_compute_kernels_id,
+        num_cores_per_batch,
+        is_output_sharded,
+        cb_out4_id
+        ]
+    (
+        const void* operation,
+        Program& program,
+        const std::vector<Tensor>& input_tensors,
+        const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+        const std::vector<Tensor>& output_tensors
+    ) {
+
+        auto q_buffer = input_tensors.at(0).buffer();
+        auto k_buffer = input_tensors.at(1).buffer();
+        auto v_buffer = input_tensors.at(2).buffer();
+        auto mask_buffer = optional_input_tensors.at(0).has_value() ? optional_input_tensors.at(0).value().buffer() : nullptr;
+        TT_FATAL(mask_buffer != nullptr);
+
+        auto out0_buffer = output_tensors.at(0).buffer();
+        uint32_t q_addr = q_buffer->address();
+        uint32_t k_addr = k_buffer->address();
+        uint32_t v_addr = v_buffer->address();
+        uint32_t mask_addr = mask_buffer->address();
+        uint32_t out_addr = out0_buffer->address();
+
+        // Set reader rt args
+        for (uint32_t i = 0; i < num_active_cores; ++i) {
+            CoreCoord core = core_group[i];
+            uint32_t worker_id = i % num_cores_per_batch - 1;
+            bool do_reduce = (worker_id == -1);
+
+            uintptr_t reader_kernels_id = all_reader_kernels_id[i];
+            uintptr_t writer_kernels_id = all_writer_kernels_id[i];
+            uintptr_t compute_kernels_id = all_compute_kernels_id[i];
+
+            SetRuntimeArgs(program, reader_kernels_id, core, { q_addr, k_addr, v_addr, mask_addr });
+            if (do_reduce) {
+                SetRuntimeArgs(program, writer_kernels_id, core, { out_addr });
+            }
+        }
+
+        if (is_output_sharded) {
```

Comment:
Did I miss where override_runtime_args updates the address of Q input if it is sharded?

Reply:
Is that unnecessary because q_addr is passed as runtime arg and it's explicitly read from remote cores?

Reply:
Are you referring to `UpdateDynamicCircularBufferAddress`?

I did not have it for Q because I did not assign Q to a cb of the reducer core

Reply:
Thanks, that's what I meant. I realize it's not necessary because of what you said

Reply:
sounds good! I will mark it as resolved